### PR TITLE
Oscilloscope and Multivibrator

### PR DESCRIPTION
**Pull request by ragv**
<hr>

**SUMMARY OF ADDITIONS / MODIFICATIONS**

<hr>

**1. Added `Multivibrator` - Uses threading concept**
A Triple mode ( Monoshot, Astable, and bistable ) multivibrator.

**2. Added a text based `Oscilloscope` - Uses threading concept**

Added a text based oscilloscope which can help users view the time waveforms of connected elements in the terminal. 

Features include setting up of the scale to suit the users needs and a `setColour()` method to set the display colour [ only in supported operating systems ] of the oscilloscope. 

The oscilloscope module can be reused without reinitialization for displaying a different waveform.

**3. Enhanced the output of `drawIC()` method to make the vertical lines look connected and the notch in the IC more natural.**

**4. Changed connector class to include `isInputof` `isOutputof` and `untap` methods.**
Connection checking and untap method added.
USAGE:

```
>>>    a = Connector(1); b= Connector(1); o = Connector();
>>>    g1 = AND(a,b)
>>>    g1.setOutput(o)
>>>    a.isInputof(g1)
True
>>>    a.untap(g1)
>>>    a.isInputof(g1)
False
>>>    o.isInputof(g1), o.isOutputof(g1)
False, True
```

**5. Added Examples for Multivibrator in different modes**

**6. Modified examples of all FlipFlops and Counters to include oscilloscope preview**

**7. Corrected Exception handling statement to be compatible with both Python 2.x and Python 3.x by replacing `e` by `sys.exc_info()[1].args[0]`**

**8. `autopep8` - ed the codebase (Redid the same after Rebase too)**

<hr>

**PREVIEWS:**

<hr>

**1. Usage of Multivibrator in mode 2 with the oscilloscope:**

```
Output = Connector()

#Initiating the mulivibrator

Mult1 = Multivibrator(0, 2, on_time=0.2, off_time=0.8)
Mult1.start()

out = Connector()
Mult1.setOutput(Output)

Mult1.trigger()

#Initiating the oscilloscope

o = Oscilloscope((Output, 'OUT'))
o.start()
o.setScale(0.2)
o.unhold()
time.sleep(10)
o.hold() 
# o.hold() is not required as it is included in the display. 
# It is shown to get a clear picture of the working.
o.display()
```

```
=====================================================================================================================================================================
BinPy - Oscilloscope
=====================================================================================================================================================================
                                                                                                                                 SCALE - X-AXIS : 1 UNIT WIDTH = 0.2
=====================================================================================================================================================================
          │
          │
          │     ┌┐   ┌┐   ┌┐   ┌┐   ┌┐   ┌┐   ┌┐   ┌┐   ┌┐                                                                                                        
   OUT    │     ││   ││   ││   ││   ││   ││   ││   ││   ││                                                                                                        
          ─ ────┘└───┘└───┘└───┘└───┘└───┘└───┘└───┘└───┘└────────────────────────────────────────────────────────────────────────────────────────────────────────
          │
          │
│││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

```

**2. Sample Time waveform generated using the oscilloscope [ for `JKFlipFlop` ]**

[ In the actual command line the vertical lines appear solidly connected unlike the preview below ]

Please try running a sample program in `BinPy/examples/Sequential/` to get a exact preview of the oscilloscope.

```
=====================================================================================================================================================================
BinPy - Oscilloscope
=====================================================================================================================================================================
                                                                                                                                SCALE - X-AXIS : 1 UNIT WIDTH = 0.01
=====================================================================================================================================================================
          │
          │
          │                                 ┌────────────────┐                 ┌─────────────────┐                ┌──────────────┐                                
   CLK    │                                 │                │                 │                 │                │              │                                
          ─ ────────────────────────────────┘                └─────────────────┘                 └────────────────┘              └────────────────────────────────
          │
          │
          │
          │
          │ ┌───────────────────────────────┐                                   ┌─────────────────────────────────┐                                               
   J      │ │                               │                                   │                                 │                                               
          ─ ┘                               └───────────────────────────────────┘                                 └───────────────────────────────────────────────
          │
          │
          │
          │
          │                                 ┌─────────────────────────────────────────────────────────────────────┐                                               
   k      │                                 │                                                                     │                                               
          ─ ────────────────────────────────┘                                                                     └───────────────────────────────────────────────
          │
          │
          │
          │
          │ ┌────────────────────────────────────────────────┐                                   ┌──────────────────────────────────────────────────┐             
   OUT    │ │                                                │                                   │                                                  │             
          ─ ┘                                                └───────────────────────────────────┘                                                  └─────────────
          │
          │
          │
          │
          │                                                  ┌───────────────────────────────────┐                                                                
   OUT!   │                                                  │                                   │                                                                
          ─ ─────────────────────────────────────────────────┘                                   └────────────────────────────────────────────────────────────────
          │
          │
          │
          │
          │ ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐             
   ENABL  │ │                                                                                                                                       │             
          ─ ┘                                                                                                                                       └─────────────
          │
          `│
│││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││││
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

```

**6. New Look of drawIC()**

[Please note that all vertical lines will be connected - Markdown prevents it from looking connected in the preview below]

```
              ┌─────────◡─────────┐
              │                   │
     [1]    ──┤  1            14  ├──    [1]    
              │                   │
              │                   │
     [0]    ──┤  2      7     13  ├──    [0]    
              │                   │
              │                   │
     [1]    ──┤  3      4     12  ├──    [0]    
              │                   │
              │                   │
     [0]    ──┤  4      0     11  ├──    [1]    
              │                   │
              │                   │
     [0]    ──┤  5      0     10  ├──    [1]    
              │                   │
              │                   │
     [1]    ──┤  6             9  ├──    [1]    
              │                   │
              │                   │
     [0]    ──┤  7             8  ├──    [0]    
              │                   │
              └───────────────────┘  

```

### CODE REVIEW

**Comment by ghost** [comment]
@jayrambhia Have rebased to reflect changes in pr 162 ... Pl review...

**Comment by jayrambhia** [comment]
Fill in the captcha. Checking if human! How do you get so much time?

**Comment by ghost** [comment]
@jayrambhia haha! :D I just luv coding so much ... !!

File: BinPy/tools/multivibrator.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            off_time=None):
+
+        threading.Thread.__init__(self)
+
+        if frequency is not None:
+            self.time_period = 1.0 / frequency
+        if time_period is not None:
+            self.time_period = time_period
+        if time_period is None and frequency is None:
+            self.time_period = 10
```
Comment (debugger22) [inline BinPy/tools/multivibrator.py:52]: @rvraghav93 You've mentioned default `time_period = 1s` on line no. 24 but it is 10 on line no 52.

Comment (ragv) [inline BinPy/tools/multivibrator.py:52]: corrected time_period in pr #169

**Pull request merged**